File: java/src/main/java/com/alxarafe/app/domain/catalogue/entity/Item.java
package com.alxarafe.app.domain.catalogue.entity;

import com.alxarafe.app.domain.catalogue.valueobject.ItemFamilyId;
import com.alxarafe.app.domain.catalogue.valueobject.ItemId;
import com.alxarafe.app.domain.catalogue.valueobject.ItemUomConversion;
import com.alxarafe.app.domain.catalogue.valueobject.Money;
import com.alxarafe.app.domain.catalogue.valueobject.Sku;
import com.alxarafe.app.domain.catalogue.valueobject.UomId;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * Aggregate Root representing a product / merchandise item.
 */
public final class Item {

    private final ItemId id;
    private final Sku sku;
    private final String name;
    private final ItemFamilyId familyId;
    private final UomId baseUomId;
    private final boolean isBatchManaged;
    private final boolean isExpirable;
    private final Money baseCost;
    private final List<ItemUomConversion> uomConversions;

    public Item(ItemId id, Sku sku, String name, ItemFamilyId familyId,
                UomId baseUomId, boolean isBatchManaged, boolean isExpirable,
                Money baseCost) {
        this.id = Objects.requireNonNull(id, "Item id cannot be null.");
        this.sku = Objects.requireNonNull(sku, "Item sku cannot be null.");
        Objects.requireNonNull(name, "Item name cannot be null.");
        if (name.isBlank()) {
            throw new IllegalArgumentException("Item name cannot be empty.");
        }
        this.name = name;
        this.familyId = Objects.requireNonNull(familyId, "Item familyId cannot be null.");
        this.baseUomId = Objects.requireNonNull(baseUomId, "Item baseUomId cannot be null.");
        if (isExpirable && !isBatchManaged) {
            throw new IllegalArgumentException(
                    "An expirable item must also be batch-managed.");
        }
        this.isBatchManaged = isBatchManaged;
        this.isExpirable = isExpirable;
        this.baseCost = Objects.requireNonNull(baseCost, "Item baseCost cannot be null.");
        this.uomConversions = new ArrayList<>();
    }

    public ItemId id() {
        return id;
    }

    public Sku sku() {
        return sku;
    }

    public String name() {
        return name;
    }

    public ItemFamilyId familyId() {
        return familyId;
    }

    public UomId baseUomId() {
        return baseUomId;
    }

    public boolean isBatchManaged() {
        return isBatchManaged;
    }

    public boolean isExpirable() {
        return isExpirable;
    }

    public Money baseCost() {
        return baseCost;
    }

    public List<ItemUomConversion> uomConversions() {
        return Collections.unmodifiableList(uomConversions);
    }

    /**
     * Adds a UoM conversion to this item.
     *
     * @throws IllegalArgumentException if a duplicate conversion exists
     */
    public void addUomConversion(ItemUomConversion conversion) {
        Objects.requireNonNull(conversion, "UoM conversion cannot be null.");
        boolean duplicate = uomConversions.stream().anyMatch(existing ->
                existing.fromUomId().equals(conversion.fromUomId())
                        && existing.toUomId().equals(conversion.toUomId()));
        if (duplicate) {
            throw new IllegalArgumentException("Duplicate UoM conversion for this item.");
        }
        uomConversions.add(conversion);
    }
}

File: php/src/Domain/Catalogue/Entity/Item.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Catalogue\Entity;

use Alxarafe\App\Domain\Catalogue\ValueObject\ItemFamilyId;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemId;
use Alxarafe\App\Domain\Catalogue\ValueObject\ItemUomConversion;
use Alxarafe\App\Domain\Catalogue\ValueObject\Sku;
use Alxarafe\App\Domain\Catalogue\ValueObject\UomId;
use InvalidArgumentException;

/**
 * Aggregate Root representing a product / merchandise item.
 */
final class Item
{
    /** @var ItemUomConversion[] */
    private array $uomConversions = [];

    public function __construct(
        private readonly ItemId $id,
        private readonly Sku $sku,
        private readonly string $name,
        private readonly ItemFamilyId $familyId,
        private readonly UomId $baseUomId,
        private readonly bool $isBatchManaged,
        private readonly bool $isExpirable,
    ) {
        if ($name === '') {
            throw new InvalidArgumentException('Item name cannot be empty.');
        }

        if ($isExpirable && !$isBatchManaged) {
            throw new InvalidArgumentException(
                'An expirable item must also be batch-managed.'
            );
        }
    }

    public function id(): ItemId
    {
        return $this->id;
    }

    public function sku(): Sku
    {
        return $this->sku;
    }

    public function name(): string
    {
        return $this->name;
    }

    public function familyId(): ItemFamilyId
    {
        return $this->familyId;
    }

    public function baseUomId(): UomId
    {
        return $this->baseUomId;
    }

    public function isBatchManaged(): bool
    {
        return $this->isBatchManaged;
    }

    public function isExpirable(): bool
    {
        return $this->isExpirable;
    }

    /**
     * @return ItemUomConversion[]
     */
    public function uomConversions(): array
    {
        return $this->uomConversions;
    }

    /**
     * Adds a UoM conversion to this item.
     *
     * @throws InvalidArgumentException if a duplicate conversion exists
     */
    public function addUomConversion(ItemUomConversion $conversion): void
    {
        foreach ($this->uomConversions as $existing) {
            if (
                $existing->fromUomId()->equals($conversion->fromUomId())
                && $existing->toUomId()->equals($conversion->toUomId())
            ) {
                throw new InvalidArgumentException(
                    'Duplicate UoM conversion for this item.'
                );
            }
        }

        $this->uomConversions[] = $conversion;
    }
}
